### ISSUE TITLE
mathjax and AsciiMath

### ISSUE THREAD

**Issue by guillaumecherel** [jgm/pandoc#2654]
I want to mathjax with AsciiMath. This means using the option `--mathjax="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=AM_HTMLorMML"`.

With AsciiMath, math formulas must be enclosed between `...` in the html file. In the markdown document, I escape the backticks to produce the desired html output.

I noticed that if nothing of the form `$...$` appears in the markdown document, pandoc doesn't include the script element in the html which loads the mathjax script. As a result, my AsciiMath formulas were displayed as plain text and not rendered as expected.

So, in order to use AsciiMath, it's necessary to add a `$...$` formula somewhere in your document.

Perhaps a fix would be to always include the script element whenever the --mathjax option is specified?

**Comment by jgm** [comment]
Here's a diff that would do the job.  But I'm not persuaded we should.

The advantage of keeping track of whether math is used is that authors can use the same options for converting a bunch of files (e.g. in a blog) without worrying about which of them contain math.  When documents don't have math, you don't pay the cost of a javascript load.

```diff
diff --git a/src/Text/Pandoc/Writers/HTML.hs b/src/Text/Pandoc/Writers/HTML.hs
index 73a8906..505d90d 100644
--- a/src/Text/Pandoc/Writers/HTML.hs
+++ b/src/Text/Pandoc/Writers/HTML.hs
@@ -72,7 +72,6 @@ import Data.Aeson (Value)
 
 data WriterState = WriterState
     { stNotes            :: [Html]  -- ^ List of notes
-    , stMath             :: Bool    -- ^ Math is used in document
     , stQuotes           :: Bool    -- ^ <q> tag is used
     , stHighlighting     :: Bool    -- ^ Syntax highlighting is used
     , stSecNum           :: [Int]   -- ^ Number of current section
@@ -80,7 +79,7 @@ data WriterState = WriterState
     }
 
 defaultWriterState :: WriterState
-defaultWriterState = WriterState {stNotes= [], stMath = False, stQuotes = False,
+defaultWriterState = WriterState {stNotes= [], stQuotes = False,
                                   stHighlighting = False, stSecNum = [],
                                   stElement = False}
 
@@ -140,47 +139,43 @@ pandocToHtml opts (Pandoc meta blocks) = do
   st <- get
   let notes = reverse (stNotes st)
   let thebody = blocks' >> footnoteSection opts notes
-  let  math = if stMath st
-                then case writerHTMLMathMethod opts of
-                           LaTeXMathML (Just url) ->
-                              H.script ! A.src (toValue url)
-                                       ! A.type_ "text/javascript"
-                                       $ mempty
-                           MathML (Just url) ->
-                              H.script ! A.src (toValue url)
-                                       ! A.type_ "text/javascript"
-                                       $ mempty
-                           MathJax url ->
-                              H.script ! A.src (toValue url)
-                                       ! A.type_ "text/javascript"
-                                       $ case writerSlideVariant opts of
-                                              SlideousSlides ->
-                                                 preEscapedString
-                                                 "MathJax.Hub.Queue([\"Typeset\",MathJax.Hub]);"
-                                              _ -> mempty
-                           JsMath (Just url) ->
-                              H.script ! A.src (toValue url)
-                                       ! A.type_ "text/javascript"
-                                       $ mempty
-                           KaTeX js css ->
-                              (H.script ! A.src (toValue js) $ mempty) <>
-                              (H.link ! A.rel "stylesheet" ! A.href (toValue css)) <>
-                              (H.script ! A.type_ "text/javascript" $ toHtml renderKaTeX)
-                           _ -> case lookup "mathml-script" (writerVariables opts) of
-                                      Just s | not (writerHtml5 opts) ->
-                                        H.script ! A.type_ "text/javascript"
-                                           $ preEscapedString
-                                            ("/*<![CDATA[*/\n" ++ s ++ "/*]]>*/\n")
-                                             | otherwise -> mempty
-                                      Nothing -> mempty
-                else mempty
+  let  math = case writerHTMLMathMethod opts of
+                      LaTeXMathML (Just url) ->
+                         H.script ! A.src (toValue url)
+                                  ! A.type_ "text/javascript"
+                                  $ mempty
+                      MathML (Just url) ->
+                         H.script ! A.src (toValue url)
+                                  ! A.type_ "text/javascript"
+                                  $ mempty
+                      MathJax url ->
+                         H.script ! A.src (toValue url)
+                                  ! A.type_ "text/javascript"
+                                  $ case writerSlideVariant opts of
+                                         SlideousSlides ->
+                                            preEscapedString
+                                            "MathJax.Hub.Queue([\"Typeset\",MathJax.Hub]);"
+                                         _ -> mempty
+                      JsMath (Just url) ->
+                         H.script ! A.src (toValue url)
+                                  ! A.type_ "text/javascript"
+                                  $ mempty
+                      KaTeX js css ->
+                         (H.script ! A.src (toValue js) $ mempty) <>
+                         (H.link ! A.rel "stylesheet" ! A.href (toValue css)) <>
+                         (H.script ! A.type_ "text/javascript" $ toHtml renderKaTeX)
+                      _ -> case lookup "mathml-script" (writerVariables opts) of
+                                 Just s | not (writerHtml5 opts) ->
+                                   H.script ! A.type_ "text/javascript"
+                                      $ preEscapedString
+                                       ("/*<![CDATA[*/\n" ++ s ++ "/*]]>*/\n")
+                                        | otherwise -> mempty
+                                 Nothing -> mempty
   let context =   (if stHighlighting st
                       then defField "highlighting-css"
                              (styleToCss $ writerHighlightStyle opts)
                       else id) $
-                  (if stMath st
-                      then defField "math" (renderHtml math)
-                      else id) $
+                  defField "math" (renderHtml math) $
                   defField "quotes" (stQuotes st) $
                   maybe id (defField "toc" . renderHtml) toc $
                   defField "author-meta" authsMeta $
@@ -487,8 +482,7 @@ blockToHtml opts (RawBlock f str)
   | f == Format "html" = return $ preEscapedString str
   | f == Format "latex" =
       case writerHTMLMathMethod opts of
-           MathJax _  -> do modify (\st -> st{ stMath = True })
-                            return $ toHtml str
+           MathJax _  -> return $ toHtml str
            _          -> return mempty
   | otherwise          = return mempty
 blockToHtml opts (HorizontalRule) = return $ if writerHtml5 opts then H5.hr else H.hr
@@ -751,7 +745,6 @@ inlineToHtml opts inline =
                                else (\x -> leftQuote >> x >> rightQuote)
                                     `fmap` inlineListToHtml opts lst
     (Math t str) -> do
-      modify (\st -> st {stMath = True})
       let mathClass = toValue $ ("math " :: String) ++
                       if t == InlineMath then "inline" else "display"
       case writerHTMLMathMethod opts of
@@ -812,10 +805,8 @@ inlineToHtml opts inline =
     (RawInline f str)
       | f == Format "latex" ->
                           case writerHTMLMathMethod opts of
-                               LaTeXMathML _ -> do modify (\st -> st {stMath = True})
-                                                   return $ toHtml str
-                               MathJax _     -> do modify (\st -> st {stMath = True})
-                                                   return $ toHtml str
+                               LaTeXMathML _ -> return $ toHtml str
+                               MathJax _     -> return $ toHtml str
                                _             -> return mempty
       | f == Format "html" -> return $ preEscapedString str
       | otherwise          -> return mempty
```

**Comment by guillaumecherel** [comment]
I see. Another possibility would be to allow backticks as an additional syntax for math, but I suppose it would conflict with the current use of backticks. Maybe escaped backticks then? That wouldn't have to do anything special, since escaped backticks already get turned into simple backticks in the resulting html, but it would add the script html element.

**Comment by jgm** [comment]
You could just add the script element to your template, or
use --include-in-header.

+++ guillaumecherel [Jan 15 16 01:39 ]:
